src/index/CompressedRelation.cpp

```diff
@@ -312,80 +341,140 @@ bool CompressedRelationReader::FilterDuplicatesAndGraphs::canBlockBeSkipped(
 }
 
 // _____________________________________________________________________________
-CompressedRelationReader::IdTableGenerator CompressedRelationReader::lazyScan(
+CompressedRelationReader::IdTableGeneratorInputRange
+CompressedRelationReader::lazyScan(
     ScanSpecification scanSpec,
     std::vector<CompressedBlockMetadata> relevantBlockMetadata,
     ColumnIndices additionalColumns, CancellationHandle cancellationHandle,
     const LocatedTriplesPerBlock& locatedTriplesPerBlock,
     LimitOffsetClause limitOffset) const {
   AD_CONTRACT_CHECK(cancellationHandle);
 
-  // We will modify `limitOffset` as we go. We make a copy of the original
-  // value for some sanity checks at the end of the function.
-  const auto originalLimit = limitOffset;
-
   if (relevantBlockMetadata.empty()) {
-    co_return;
+    return IdTableGeneratorInputRange{};
   }
 
-  // Some preparation.
-  auto [beginBlockMetadata, endBlockMetadata] =
-      getBeginAndEnd(relevantBlockMetadata);
-  LazyScanMetadata& details = co_await cppcoro::getDetails;
-  size_t numBlocksTotal = endBlockMetadata - beginBlockMetadata;
-  auto config =
-      getScanConfig(scanSpec, additionalColumns, locatedTriplesPerBlock);
+  struct Generator : ad_utility::InputRangeFromGet<IdTable, LazyScanMetadata> {
+    ScanSpecification scanSpec;
+    std::vector<CompressedBlockMetadata> relevantBlockMetadata;
+    ColumnIndices additionalColumns;
+    CancellationHandle cancellationHandle;
+    const LocatedTriplesPerBlock& locatedTriplesPerBlock;
+    LimitOffsetClause limitOffset;
+    ad_utility::InputRangeTypeErased<IdTable, LazyScanMetadata>
+        blockGenerator{};
+    bool firstBlockYielded{false};
+    bool middleBlocksYielded{false};
+    bool lastBlockYielded{false};
+    bool generatorCreated{false};
+    std::vector<CompressedBlockMetadata>::iterator beginBlockMetadata;
+    std::vector<CompressedBlockMetadata>::iterator endBlockMetadata;
+    const CompressedRelationReader* reader;
+    ScanImplConfig config;
+    IdTableGeneratorInputRange middleBlocksGenerator{};
+    // We will modify `limitOffset` as we go. We make a copy of the original
+    // value for some sanity checks at the end of the function.
+    const LimitOffsetClause originalLimit{limitOffset};
+    std::size_t numBlocksTotal;
```

Comment:
If you rearrange your members to have the dependent members at the end (the iterators + num blocks total) and then 
directly initialize them on the declaration, then you can get rid of the constructor, because struct initialization will do the right thing for you.